File: src/week4/P001.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m, k, s] = input[0].split(' ').map(Number);
const graph = Array.from({ length: n + 1 }, () => []);

for (let i = 1; i <= m; i++) {
    const [a, b] = input[i].split(' ').map(Number);
    graph[a].push(b);
}
const dist = new Array(n + 1).fill(-1);
const queue = [];
let head = 0; // 큐의 shift() 연산(O(N))을 대체할 포인터 (O(1))

// 6. BFS 시작 노드 설정
queue.push(s);
dist[s] = 0; // 시작 도시의 거리는 0

// 7. BFS 수행
while (head < queue.length) {
    // 큐에서 요소를 꺼냅니다. (shift() 대신 head 포인터 사용)
    const now = queue[head++];

    // 현재 도시(now)에서 갈 수 있는 다음 도시(next)들을 순회합니다.
    for (const next of graph[now]) {
        // 만약 다음 도시(next)를 아직 방문하지 않았다면
        if (dist[next] === -1) {
            // 거리를 갱신하고 (현재 도시 거리 + 1)
            dist[next] = dist[now] + 1;
            // 큐에 추가합니다.
            queue.push(next);
        }
    }
}

const answer = [];
for (let i = 1; i <= n; i++) {
    // 거리가 k인 도시를 찾습니다.
    if (dist[i] === k) {
        answer.push(i);
    }
}

if (answer.length === 0) console.log(-1);
else console.log(answer.join('\n'));

File: src/week4/P001.java
package week4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class P001 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); // 도시의 개수 (노드)
        int m = Integer.parseInt(st.nextToken()); // 도로의 개수 (간선)
        int k = Integer.parseInt(st.nextToken()); // 거리 정보
        int s = Integer.parseInt(st.nextToken()); // 출발 도시

        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
        }

        int[] dist = new int[n + 1];
        Arrays.fill(dist, -1);

        Queue<Integer> queue = new LinkedList<>();

        queue.add(s);
        dist[s] = 0;

        while (!queue.isEmpty()) {
            int now = queue.poll();

            for (int next : graph.get(now)) {
                // 아직 방문하지 않은 노드라면
                if (dist[next] == -1) {
                    dist[next] = dist[now] + 1; // 거리 갱신
                    queue.add(next);
                }
            }
        }
        StringBuilder sb = new StringBuilder();
        boolean found = false;
        for (int i = 1; i <= n; i++) {
            if (dist[i] == k) {
                sb.append(i).append("\n");
                found = true;
            }
        }

        if (!found) System.out.println(-1);
        else System.out.print(sb.toString());
    }
}
